### DISCUSSION TITLE
Integrating Prometheus with NextJS?

### DISCUSSION THREAD

**Discussion by RyanPayso13** [vercel/next.js#24980] [Help]
We have a project requirement to integrate [Prometheus](https://prometheus.io/) with NextJS.

We are currently unsure as to **_how_** it would be possible to integrate Prometheus into the NodeJS layer of NextJS.

We found this issue which claims to have gotten the two integrated:

https://github.com/vercel/next.js/discussions/16205

But when we tried with a custom `server.js` at the project root it seemed to be completely ignored in either dev mode or with a `build` and `start` of the server.

Has anyone got anymore experience integrating these two or any other metrics reporting tools at this level with NextJS?

**Comment by weyert** [comment]
Yes, I have been successfully been using Next.js custom server with  Prometheus metrics via Opentelemetry-js Metrics SDK. The only problem is that I can't find a way to calculate the requested url back into a route path (e.g. `/users/[id]/index.tsx`) meaning I have high cardinality metrics wise.

Now that I am writing this I think I might be able to somehow monkey patch Next.js server in such manner that I might be able to extract it somehow and store it in a async context so I can refer to it in the Express metrics middleware
